modules/pam_env/tst-pam_env-retval.c

```diff
@@ -0,0 +1,273 @@
+
+static const char user_name[] = "";
+static struct pam_conv conv;
+
+static void setup(void)
+{
+  FILE *fp;
+
+  ASSERT_EQ(0, mkdir(tests, 0755));
+  ASSERT_EQ(0, mkdir(tests_usr, 0755));
+  ASSERT_EQ(0, mkdir(tests_usr_etc, 0755));
+  ASSERT_EQ(0, mkdir(tests_usr_etc_security, 0755));
+
+  ASSERT_NE(NULL, fp = fopen(my_pam_env_conf, "w"));
+  ASSERT_LT(0, fprintf(fp,
+		       "PAGER		DEFAULT=vi\n"
+		       "MANPAGER	DEFAULT=vim\n"));
+  ASSERT_EQ(0, fclose(fp));
+  ASSERT_NE(NULL, fp = fopen(my_environment, "w"));
+  ASSERT_LT(0, fprintf(fp,
+	               "test_value=foo\n"
+		       "test2_value=bar\n"));
+  ASSERT_EQ(0, fclose(fp));
+  ASSERT_NE(NULL, fp = fopen(environment, "w"));
+  ASSERT_LT(0, fprintf(fp,
+	               "var_etc_test=foo\n"
+		       "var_etc_test2=bar\n"));
+  ASSERT_EQ(0, fclose(fp));
+  ASSERT_NE(NULL, fp = fopen(pam_env_conf, "w"));
+  ASSERT_LT(0, fprintf(fp,
+		       "PAGER		DEFAULT=emacs\n"
+		       "MANPAGER	DEFAULT=less\n"));
+  ASSERT_EQ(0, fclose(fp));
+}
+
+static void cleanup(void)
+{
+  ASSERT_EQ(0, unlink(my_pam_env_conf));
+  ASSERT_EQ(0, unlink(my_environment));
+  ASSERT_EQ(0, unlink(environment));
+  ASSERT_EQ(0, unlink(pam_env_conf));
+  ASSERT_EQ(0, rmdir(tests_usr_etc_security));
+  ASSERT_EQ(0, rmdir(tests_usr_etc));
+  ASSERT_EQ(0, rmdir(tests_usr));
+  ASSERT_EQ(0, rmdir(tests));
+}
+
+static void check_array(const char **list1, const size_t list1_length, char **list2,  const size_t list2_length)
+{
+  ASSERT_LT(1,(int)list1_length);
+
+  for (size_t i = 0; i < list1_length; i++  ) {
+    int found = 0;
+    for (size_t j = 0; j < list2_length; j++  )
+      if (strcmp(list1[i], list2[j]) == 0)
+	found = 1;
+    ASSERT_EQ(1, found);
+  }
+}
+
+static void check_env(const char **compare_list,
+		      const size_t compare_list_length,
+		      pam_handle_t *pamh) {
+  ASSERT_NE(NULL, pamh);
+
+  ASSERT_EQ(PAM_SUCCESS, pam_authenticate(pamh, PAM_SILENT));
+  ASSERT_EQ(PAM_SUCCESS, pam_setcred(pamh, PAM_SILENT));
+  ASSERT_EQ(PAM_SUCCESS, pam_open_session(pamh, PAM_SILENT));
+
+  char **pam_envlist = pam_getenvlist(pamh);
+  if (pam_envlist != NULL) {
+    size_t list2_length = 0;
+    for (char **x = pam_envlist; *x != NULL; ++x) list2_length++;
+
+    check_array(compare_list, compare_list_length,
+		pam_envlist, list2_length);
+    for (char **pam_env = pam_envlist; *pam_env != NULL; ++pam_env) {
+      free(*pam_env);
+    }
+    free(pam_envlist);
+  }
+}
+
+
+int
+main(void)
+{
+  pam_handle_t *pamh = NULL;
+  FILE *fp;
+  char cwd[PATH_MAX];
+
+  ASSERT_NE(NULL, getcwd(cwd, sizeof(cwd)));
+
+  setup();
+
+  /* PAM_SUCCESS -> PAM_SUCCESS, PAM_IGNORE -> PAM_PERM_DENIED */
+  ASSERT_NE(NULL, fp = fopen(service_file, "w"));
+  ASSERT_LT(0, fprintf(fp, "#%%PAM-1.0\n"
+		       "auth required %s/.libs/%s.so\n"
+		       "account required %s/.libs/%s.so\n"
+		       "password required %s/.libs/%s.so\n"
+		       "session required %s/.libs/%s.so\n",
```

Comment:
This is undefined behaviour: `pam_env` returns `PAM_SUCCESS` when environment variables are set.
Please have a look at [ldv/tst-pam_env-retval](https://github.com/linux-pam/linux-pam/commit/b53b9511274d2d98b2055a9b4ab7e635d3c9727a).